com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java

```diff
@@ -0,0 +1,157 @@
+        return "com.microsoft.java.debug.AdvancedLaunchingConnector";
+    }
+
+    @Override
+    public VirtualMachine launch(Map<String, ? extends Argument> connectionArgs)
+            throws IOException, IllegalConnectorArgumentsException, VMStartException {
+        String cwd = connectionArgs.get(CWD).value();
+        File workingDir = null;
+        if (cwd != null && Files.isDirectory(Paths.get(cwd))) {
+            workingDir = new File(cwd);
+        }
+
+        String[] envVars = null;
+        try {
+            envVars = DebugUtility.decodeArrayArgument(connectionArgs.get(ENV).value());
+        } catch (IllegalArgumentException e) {
+            // do nothing.
+        }
+
+        SocketListeningConnectorImpl listenConnector = new SocketListeningConnectorImpl(
+                virtualMachineManager());
+        Map<String, Connector.Argument> args = listenConnector.defaultArguments();
+        ((Connector.IntegerArgument) args.get("timeout")).setValue(ACCEPT_TIMEOUT); //$NON-NLS-1$
+        String address = listenConnector.startListening(args);
+
+        String[] cmds = constructLaunchCommand(connectionArgs, address);
+        Process process = Runtime.getRuntime().exec(cmds, envVars, workingDir);
+
+        VirtualMachineImpl vm;
+        try {
+            vm = (VirtualMachineImpl) listenConnector.accept(args);
+        } catch (IOException | IllegalConnectorArgumentsException e) {
+            process.destroy();
+            throw new VMStartException(String.format("VM did not connect within given time: %d ms", ACCEPT_TIMEOUT), process);
+        }
+
+        vm.setLaunchedProcess(process);
+        return vm;
+    }
+
+    private static String[] constructLaunchCommand(Map<String, ? extends Argument> launchingOptions, String address) {
+        String javaHome = launchingOptions.get(HOME).value();
+        String javaOptions = launchingOptions.get(OPTIONS).value();
+        String main = launchingOptions.get(MAIN).value();
+        boolean suspend = Boolean.valueOf(launchingOptions.get(SUSPEND).value());
+        String javaExec = launchingOptions.get(EXEC).value();
+        String slash = System.getProperty("file.separator");
+
+        List<String> cmds = new ArrayList<>();
+        cmds.add(javaHome + slash + "bin" + slash + javaExec);
+        cmds.addAll(parseArguments("-Xdebug -Xnoagent -Djava.compiler=NONE"));
+        cmds.add("-Xrunjdwp:transport=dt_socket,address=" + address + ",server=n,suspend=" + (suspend ? "y" : "n"));
+        cmds.addAll(parseArguments(javaOptions));
+        cmds.addAll(parseArguments(main));
+
+        return cmds.toArray(new String[0]);
+    }
+
+    /**
+     * Parses the given command line into separate arguments that can be passed
+     * to <code>Runtime.getRuntime().exec(cmdArray)</code>.
+     *
+     * @param args command line as a single string.
+     * @return the arguments array.
+     */
+    private static List<String> parseArguments(String cmdStr) {
```

Comment:
Can we use the implementation from the base class?

Reply:
fixed.